### ISSUE TITLE
Conditionally applying policies

### ISSUE THREAD

**Issue by sreedhap** [eclipse/ditto#239]
Hi folks, my team currently has a use case to apply policies conditionally. 

**Scenario** 
We want to restrict access based on the value of one of the significant attributes of a thing - like their _**type**_ - _air-conditioner_ or _elevator_.

[AWS Policies]( https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition.html) implement a '_Condition_' element for a similar requirement.

1. Does Ditto have something like this (couldn't find it in documentation so far) or do you folks have any thoughts around this?

2. Is '_namespace_' attribute for a thing a good fit for including important attributes like the type of a thing, as mentioned in the example above? If so, then namespace in policy would take care of this I guess. 
:thinking:

**Comment by thjaeckle** [comment]
Hi there.
That's an interesting scenario there - can you share what your team is building with Ditto? We are always looking forward in hearing from people using Ditto.

Ditto's policies unfortunately do not have a comparable "dynamic condition" concept. Until now they purely are based on subjects and JSON hierarchy.
It would be technically **very** challenging to put such conditions in our policies (as we for example also apply the policies when using the "search" functionality - you cannot find something which you are not allowed to 'READ').

I would however suggest another approach in solving your problem (using the JSON hierarchy).
Model your Things with features `air-conditioner` and `elevator`, e.g.:
```json
{
  "thingId": "org.eclipse.ditto.example:aircond1",
  "policyId": "org.eclipse.ditto.example:common-policy",
  "features": {
    "air-conditioner": {
      "properties": {
        "room-temperature": 25.7,
        "desired-temperature": 22,
        "cooling": true
      }
    }
  }
}

{
  "thingId": "org.eclipse.ditto.example:elevator4711",
  "policyId": "org.eclipse.ditto.example:common-policy",
  "features": {
    "elevator": {
      "properties": {
        "current-floor": 2,
        "next-floor": 5,
        "current-speed": 45.7
      }
    }
  }
}
```

That way you can have the same policy the two things both use, e.g.:
```json
{
  "policyId": "org.eclipse.ditto.example:common-policy",
  "entries": {
    "admin": {
      "subjects": {
        "nginx:admin": {
          "type": "nginx basic auth user: may modify policy + read and write all Things"
        }
      },
      "resources": {
        "thing:/": {
          "grant": ["READ", "WRITE"],
          "revoke": []
        },
        "policy:/": {
          "grant": ["READ", "WRITE"],
          "revoke": []
        },
        "message:/": {
          "grant": ["READ", "WRITE"],
          "revoke": []
        }
      }
    },
    "air-conditioner-user": {
        "subjects": {
          "nginx:air-cond-user": {
            "type": "nginx basic auth user - may read+write feature air-conditioner"
          },
          "nginx:air-cond-user-2": {
            "type": "nginx basic auth user - may read+write feature air-conditioner"
          }
        },
        "resources": {
          "thing:/features/air-conditioner": {
            "grant": ["READ", "WRITE"],
            "revoke": []
          }
        }
      },
      "elevator-user": {
        "subjects": {
          "nginx:elevator-user": {
            "type": "nginx basic auth user - may read+write feature elevator"
          }
        },
        "resources": {
          "thing:/features/air-conditioner": {
            "grant": ["READ", "WRITE"],
            "revoke": []
          }
        }
      }
    }
  }
}
```

So in a nutshell: Use the "type" information as features in order to be able to use the resulting structure in your policy.

Regarding your second question: no, namespace is not a good place to put such information as that is nothing the policy looks at when checking access.

I hope that helps :)

**Comment by sreedhap** [comment]
Hi @thjaeckle , thanks a lot for the detailed answer. 
We aren't using Ditto at the moment, we already have a few homegrown services, and we'd like to align with Ditto specifications, so we can make a switch to Ditto at some point when we have the capacity. 

Regarding your suggestion, feature looks like a good place to model the type of thing. But this implementation only restricts the user at this feature level, what if we don't want to give the user access to whole thing including top level meta attributes, so they would not show up in the search results at all?

Thanks,
Sree & @jriehn pairing.

**Comment by thjaeckle** [comment]
closing due to inactivity to question - feel free to ask back when you get the chance to look at it :)
